### PR TITLE
Use atomics to avoid locking instructions in hot path.

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR

### CODE REVIEW

File: messagebus/src/vespa/messagebus/network/rpctarget.cpp [hunk 1]
```diff
@@ -25,11 +25,12 @@ RPCTarget::~RPCTarget()
...
 {
-    bool hasVersion = false;
     bool shouldInvoke = false;
-    {
+    ResolveState state = _state.load(std::memory_order_relaxed);
+    bool hasVersion = (state == VERSION_RESOLVED);
+    if ( ! hasVersion ) {
         vespalib::MonitorGuard guard(_lock);
-        if (_state == VERSION_RESOLVED || _state == PROCESSING_HANDLERS) {
+        if (state == PROCESSING_HANDLERS) {
```
Comment (vekterli) [inline messagebus/src/vespa/messagebus/network/rpctarget.cpp]: By removing the check for `_state == VERSION_RESOLVED`, doesn't this open up for the following race condition?
1. Thread reads `_state` == `TARGET_INVOKED` cached into local `state` variable, attempts to enter critical section. Is put on hold by kernel.
1. MBus thread receives resolved version, enters critical section, finishes processing handlers and finally sets `_state = VERSION_RESOLVED`
1. Original thread wakes up, enters critical section
1. Original thread sees that cached `state` value is _not_ equal to `PROCESSING_HANDLERS`, does not enter branch that can set `hasVersion = true`
1. Original thread initiates new version fetch towards peer even though version has already been resolved
1. New round of same race condition possible by any other subsequent concurrent senders

Note that it should be safe to use the original expression with both loads being made explicitly relaxed, as visibility should be guaranteed from entering the critical section. That should avoid the race condition while not introducing any memory fences.

**Review by vekterli — COMMENTED** [review]
See comments regarding some possible memory visibility edge cases

File: messagebus/src/vespa/messagebus/network/rpctarget.cpp [hunk 2]
```diff
@@ -25,11 +25,12 @@ RPCTarget::~RPCTarget()
 void
 RPCTarget::resolveVersion(duration timeout, RPCTarget::IVersionHandler &handler)
 {
-    bool hasVersion = false;
     bool shouldInvoke = false;
-    {
+    ResolveState state = _state.load(std::memory_order_relaxed);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpctarget.cpp]: Fixed

File: messagebus/src/vespa/messagebus/network/rpctarget.cpp [hunk 3]
```diff
@@ -54,11 +55,11 @@ RPCTarget::resolveVersion(duration timeout, RPCTarget::IVersionHandler &handler)
 bool
 RPCTarget::isValid() const
 {
-    vespalib::MonitorGuard guard(_lock);
     if (_target.IsValid()) {
         return true;
     }
-    if (_state == TARGET_INVOKED || _state == PROCESSING_HANDLERS) {
+    ResolveState state = _state.load(std::memory_order_relaxed);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpctarget.cpp]: No, it is just used once to detect if connection should be dropped and remove from pool.

File: messagebus/src/vespa/messagebus/network/rpctarget.cpp (hunk 1 above)
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpctarget.cpp]: Very good catch - Classic error.

**Pull request merged**